include/deal.II/lac/affine_constraints.templates.h

```diff
@@ -712,101 +712,150 @@ AffineConstraints<number>::close()
+                  has_sub_constraints = true;
+                }
+              else
+                sub_constraints[entry] = nullptr;
+
+            // If none of the entries in the current line refer to DoFs that are
+            // themselves constrained, then we can move on:
+            if (has_sub_constraints == false)
               {
-                // ok, this entry is further constrained:
-                chained_constraint_replaced = true;
-
-                // look up the chain of constraints for this entry
-                const size_type dof_index = line.entries[entry].first;
-                const number    weight    = line.entries[entry].second;
-
-                Assert(dof_index != line.index,
-                       ExcMessage("Cycle in constraints detected!"));
-
-                const ConstraintLine &constrained_line =
-                  lines[lines_cache[calculate_line_index(dof_index)]];
-                Assert(constrained_line.index == dof_index, ExcInternalError());
-
-                // now we have to replace an entry by its expansion. we do
-                // that by overwriting the entry by the first entry of the
-                // expansion and adding the remaining ones to the end,
-                // where we will later process them once more
-                //
-                // we can of course only do that if the DoF that we are
-                // currently handling is constrained by a linear combination
-                // of other dofs:
-                if (constrained_line.entries.size() > 0)
-                  {
-                    for (size_type i = 0; i < constrained_line.entries.size();
-                         ++i)
-                      Assert(dof_index != constrained_line.entries[i].first,
-                             ExcMessage("Cycle in constraints detected!"));
+                line_finalized[line_index] = true;
+                continue;
+              }
+
+            // Now walk through the original entries. We may replace some of
+            // them, and we may add some, but we only walk through the original
+            // entries (and also don't touch the replacements) even though in
+            // principle we could also do the ones we have added or replaced. We
+            // don't because we don't have information for those in the
+            // sub_constraints array above. But there is no harm in doing so:
+            // we will simply treat those in the next round around of the outer
+            // iteration.
+            //
+            // Since we want to delete some entries (see below) but don't want
+            // to disturb the order of the correspondence between lines.entries
+            // and sub_constraints, we store up which entries we will later
+            // have to delete.
+            const unsigned int     n_original_entries = line.entries.size();
+            std::set<unsigned int> entries_to_delete;
```

Comment:
Can we use `std::vector` (that is created outside this loop) instead? `std::set` is horrible in terms of memory allocation and not suitable for this kind of data structure. Since you insert the entries in an ordered way via the `entry` variable in the `for` loop, I see no conceptual problem, either.

Reply:
I implemented this. Just like before, the set/vector is rarely used, so I don't think it's going to make much of a difference.

Separately, though, have any of you ever looked into whether this idiom
```
std::vector<...> v;
for (some loop)
{
  v.clear();
  ...use vector...;
}
```
is in fact any faster than
```
for (some loop)
{
  std::vector<...> v;
  ...use vector...;
}
```
? In other words, does moving the declaration out of the loop make any kind of difference, or does `v.clear()` just releases all memory and resets the vector to a pristine state?

Reply:
See https://quick-bench.com/q/SAAWGNADfj_A4_q378847a0h6eQ.

Reply:
Often, the difference is bigger than so, because our sizes are between 2 and 10, or rarely 50, rather than 1000. Set the size to 10 and the difference is a factor of 3.

>  does v.clear() just releases all memory and resets the vector to a pristine state?

No, it does not release memory, at least the `libstc++` does call `erase_at_end()` here https://github.com/gcc-mirror/gcc/blob/43a0a5cd57eefd5a5bbead606ec4f6959af31802/libstdc%2B%2B-v3/include/bits/stl_vector.h#L1605-L1606 which in turn calls https://github.com/gcc-mirror/gcc/blob/master/libstdc%2B%2B-v3/include/bits/stl_vector.h#L1933-L1940. The important part is that it does not call `_M_deallocate`.

Reply:
Wow, fascinating that there is a website where you can run benchmarks!

Thank you for the education!  As mentioned, I changed the code as requested.